test/components/header_notice_test.rb

```diff
@@ -78,12 +78,12 @@ def list_with_links
 
   test "does not renders a heading if no heading is passed in" do
     render_component(branded_notice.without(:heading))
-    assert_select ".app-c-header-notice__heading", count: 0
+    assert_select "#header-notice-heading", count: 0
```

Comment:
I prefer this patten of testing using ID. Maybe even with `test-`
Part of me thinks this is worth updating like this in more places / adding notes on the gem.

Reply:
I disagree here. We generally follow BEM guidelines for our CSS and they advocate avoiding IDs - when used as selectors they cause problems with specificity and make override attempts difficult. It's not an issue in the way it's being used right here, but it opens the door to problems further down the line. I also feel it's also rarely a good idea to pollute user facing code with hooks meant solely for testing. We have a convention of the outer wrapper of our components being .app-c-thing or .gem-c-thing and every other class within cascades from that so we should be able to rely on these when it comes to the testing side.

Reply:
Hmm, having looked at where the ID is introduced I withdraw this objection! However I'm not convinced we're doing the right thing by introducing IDs like that, going to take it out for discussion elsewhere.

Reply:
> I disagree here. We generally follow BEM guidelines for our CSS and they advocate avoiding IDs - when used as selectors they cause problems with specificity and make override attempts difficult. It's not an issue in the way it's being used right here, but it opens the door to problems further down the line. 

Ah yes 100%, but _not_ for styling _only_ for testing purposes. This way the tests are independent of any gem class updates, or tag changes - they are hooked onto the element itself and are clearly prefixed.  If the `gem-` class changed to something else many tests would fail and need to be manually updated (this is happening here) Also, it's clearer to Devs at a glance updating files and tests.

Typically I see something like `[data-test]` ([rather than just ID](https://kentcdodds.com/blog/making-your-ui-tests-resilient-to-change)) being used within this context so that testing selectors and styling are separated.  If something like this was done, the Dev experience becomes simplified a testing selector is clear and the pattern is consisent.

> I also feel it's also rarely a good idea to pollute user facing code with hooks meant solely for testing. 

Hmm I'm not sure I share this concern, the extra code might be a little undesirable but potentially this could even stripped out on production at build if it's all a consistent. The benefits would be a clear consistent pattern around testing hooks that ideally work independent of gem updates.

Reply:
@maxgds just seen the update, 100% this is worthy of a FE discussion f'sure.

Reply:
Reading this thread, it reminded me of a time when I had it out a bit with Ian on an identical issue on a different PR. I think Max's point makes sense, making user-facing markup less verbose and ensuring that everything we send the user has an intent for that user is a good thing. However I do feel like this is a tricky situation because I share Chris' concern that hooking tests up to `gem-` classnames is a risk. In theory this is just a case of updating tests before you merge a dependabot PR but this seems to be a problem in govuk and a much wider discussion. I'm happy with either using the gem classnames or creating a quick `app-c` classname instead of an id.

I do think this is something worth discussing in a govuk frontenders catchup as both implementations feel imperfect. I wonder if we should be interrogating our tests and how they interact with the gem components.

Reply:
Yeah, you add 1 ID and it's a few bytes of code, but when that code gets downloaded several million times it has impact.. If we're changing the gem classes that's pretty fundamental then I'd expect tests to break and need attention - that's a feature not a bug, we ensure we haven't changed the expectations of the app when we look at what it thought it would receive. And we shouldn't be able to merge code that's breaking our tests. I don't think I want tests that don't break when the thing that's being tested fundamentally does. I could maybe be convinced with examples. I say go either way right now (although I prefer avoiding the ID) and drop this conversation onto an agenda somewhere, whether frontenders general catchup or the specific gov.uk frontenders one.

Reply:
Yeah let's pop it up for discussion, at the moment I _tend_ to see chained classes  + tags being used within testing files, so I would stick to this convention for this context of getting this PR merged.  

Maybe raise a a RFC / general topic within one of the frontend meetings??